### PR TITLE
Add documentation strings to everything

### PR DESCRIPTION
**Pull request by SchoolGuy**
This PR is there to add a lot of explanatory docs for devs who want to get into cobbler. Some strings may say that it is unclear what a parameter or method does but this will fade out the more I get into cobbler. Also this enables the automatic generated docs to be more extensive about cobbler.

### CODE REVIEW

**Comment by watologo1** [comment]
Only checked for "language syntax" issues ;)
Looks fine otherwise.

:param debug: Weather the logged message
-> Whether


for the autoinstall manager. If no logger is None
-> If logger is None (?)


 A critical message which is related to a porblem which will halt cobbler.
-> problem

:return: Weather selinux is enabled or not.
-> Whether...

**Comment by SchoolGuy** [comment]
Force push because of a rebase onto master.

**Comment by SchoolGuy** [comment]
We will definitly squash this.

File: cobbler/cobbler_collections/manager.py [hunk 1]
```diff
@@ -40,14 +40,19 @@ class CollectionManager(object):
...
+        Constructor. Manages a definitive copy of all data cobbler_collections with weakrefs
         pointing back into the class so they can understand each other's contents
         """
         self.__dict__ = CollectionManager.__shared_state
         if not CollectionManager.has_loaded:
             self.__load(api)
 
     def __load(self, api):
+        """
+        A
```
Comment (brejoc) [inline cobbler/cobbler_collections/manager.py:52]: ?? 😉

File: cobbler/cobbler_collections/manager.py [hunk 2]
```diff
@@ -65,14 +70,21 @@ def __load(self, api):
...
+        Cobbler itself does not use this GUID's though they are provided to allow for easier API linkage with other
+        applications. Cobbler uses unique names in each collection as the object id aka primary key.
+
+        :return:
         """
         return uuid.uuid4().hex
 
     def __cmp(self, a, b):
+        """
+        A
```
Comment (brejoc) [inline cobbler/cobbler_collections/manager.py:82]: ^^^

File: cobbler/cobbler_collections/manager.py [hunk 3]
```diff
@@ -187,6 +199,12 @@ def deserialize(self):
                 raise CX("serializer: error loading collection %s: %s. Check /etc/cobbler/modules.conf" % (collection.collection_type(), e))
 
     def get_items(self, collection_type):
+        """
+        A
```
Comment (brejoc) [inline cobbler/cobbler_collections/manager.py:203]: And here! ^

File: cobbler/items/item.py [hunk 4]
```diff
@@ -92,16 +92,33 @@ class Item(object):
 
     @classmethod
     def get_from_cache(cls, ref):
+        """
+        A
```
Comment (brejoc) [inline cobbler/items/item.py:96]: An other one.

File: cobbler/items/item.py [hunk 5]
```diff
@@ -92,16 +92,33 @@ class Item(object):
...
+
+        :param ref:
+        :return:
+        """
         return cls.converted_cache.get(ref.COLLECTION_TYPE, {}).get(ref.name)
 
     @classmethod
     def set_cache(cls, ref, value):
+        """
+        A
```
Comment (brejoc) [inline cobbler/items/item.py:106]: An other one.

File: cobbler/items/item.py [hunk 6]
```diff
@@ -92,16 +92,33 @@ class Item(object):
...
+        :param value:
+        """
         if ref.COLLECTION_TYPE not in cls.converted_cache:
             cls.converted_cache[ref.COLLECTION_TYPE] = {}
         cls.converted_cache[ref.COLLECTION_TYPE][ref.name] = value
 
     @classmethod
     def remove_from_cache(cls, ref):
+        """
+        A
```
Comment (brejoc) [inline cobbler/items/item.py:118]: An other one.

File: cobbler/items/item.py [hunk 7]
```diff
@@ -149,7 +166,13 @@ def __init__(self, collection_mgr, is_subobject=False):
         self.cached_dict = ""
 
     def __find_compare(self, from_search, from_obj):
+        """
+        A
```
Comment (brejoc) [inline cobbler/items/item.py:170]: An other one.

File: cobbler/modules/managers/import_signatures.py [hunk 8]
```diff
@@ -723,18 +778,31 @@ def get_repo_mirror_from_apt(self):
     # rhn-specific
 
     def rhn_repo_adder(self, distro):
-        """ not currently used """
+        """Not currently used.
```
Comment (brejoc) [inline cobbler/modules/managers/import_signatures.py:781]: Minor nitpick: The style is different from the others.

File: cobbler/modules/managers/import_signatures.py [hunk 9]
```diff
@@ -723,18 +778,31 @@ def get_repo_mirror_from_apt(self):
...
+        :param distro: Not used currently.
+        """
         return
 
     # ==========================================================================
     # rsync-specific
 
     def rsync_repo_adder(self, distro):
-        """ not currently used """
+        """Not currently used.
```
Comment (brejoc) [inline cobbler/modules/managers/import_signatures.py:791]: Minor nitpick: The style is different from the others.

File: cobbler/remote.py [hunk 10]
```diff
@@ -2228,11 +3445,24 @@ def __init__(self, args):
...
+
+        :param api: The api object to resolve information with
+        :param proxy_class: The class which proxies the requests.
+        """
         self.proxied = proxy_class(api)
         self.logger = self.proxied.api.logger
 
     def _dispatch(self, method, params, **rest):
+        """
+        A
```
Comment (brejoc) [inline cobbler/remote.py:3459]: An other one of those A's.

File: cobbler/settings.py [hunk 11]
```diff
@@ -185,11 +185,14 @@
...
+    def collection_type(self):
         """
-        return the collection name
+        This is a hardcoded string which represents the collection type.
+
+        :return: "settings"
+        :rtype: str
         """
-        return "settings"
+        return "setting"
```
Comment (brejoc) [inline cobbler/settings.py:195]: @SchoolGuy This is a change to functionality. Has this been checked by anybody?

Comment (SchoolGuy) [reply cobbler/settings.py:195]: Don't know how this could happen. I will revert this of course.

Comment (brejoc) [reply cobbler/settings.py]: 👍

File: cobbler/actions/report.py [hunk 12]
```diff
@@ -255,6 +296,12 @@ def reporting_list_names2(self, collection, name):
     def reporting_print_all_fields(self, collection, report_name, report_type, report_noheaders):
         """
         Prints all fields in a collection as a table given the report type
+
+        :param collection: The collection to report.
+        :param report_name: The name of the report.
+        :param report_type: The type of report to give.
+        :param report_noheaders: Report without the headers. (May be usefull for machine parsing)
```
Comment (hbokh) [inline cobbler/actions/report.py:303]: Fix typo: usefull --> useful (multiple times, just "grep" for it).

File: cobbler/actions/status.py [hunk 13]
```diff
@@ -44,6 +44,11 @@ class CobblerStatusReport(object):
     def __init__(self, collection_mgr, mode, logger=None):
         """
         Constructor
+
+        :param collection_mgr: The collection manager which holds all information.
+        :param mode: This describes how Cobbler should report. Currently there can be explicitly only the option
```
Comment (hbokh) [inline cobbler/actions/status.py:49]: I would rewrite this. Something like

"Currently only the option ``text`` can be set explicitly".

File: cobbler/autoinstall_manager.py [hunk 14]
```diff
@@ -225,7 +272,15 @@ def is_autoinstall_in_use(self, name):
         return False
 
     def generate_autoinstall(self, profile=None, system=None):
+        """
+        Generates the autoinstallation for a system or a profile. You may only specifify one parameter. If you specify
+        both, the system is getting generated and the profile argument is ignored.
```
Comment (hbokh) [inline cobbler/autoinstall_manager.py:277]: Might as well drop the word "getting" here.

File: cobbler/autoinstallgen.py [hunk 15]
```diff
@@ -52,6 +55,14 @@ def __init__(self, collection_mgr):
         self.templar = templar.Templar(collection_mgr)
 
     def createAutoYaSTScript(self, document, script, name):
+        """
+        This method attaches a script with a given name to an existing autoyast xml file.
+
+        :param document: The existing autoyast xml file.
```
Comment (hbokh) [inline cobbler/autoinstallgen.py:61]: SUSE calls it "AutoYaST", right? ;-)

Consistency: either use "xml" or "XML" in capital letters everywhere.

File: cobbler/autoinstallgen.py [hunk 16]
```diff
@@ -97,15 +125,14 @@ def generate_autoyast(self, profile=None, system=None, raw_data=None):
 
         document = xml.dom.minidom.parseString(raw_data)
 
-        # do we already have the #raw comment in the XML? (addComment = 0 means, don't add #raw comment)
+        # Do we already have the #raw comment in the XML? (addComment = 0 means, don't add #raw comment)
```
Comment (hbokh) [inline cobbler/autoinstallgen.py:128]: Consistency: either use "xml" or "XML" in capital letters everywhere.

File: cobbler/autoinstallgen.py [hunk 17]
```diff
@@ -146,13 +172,16 @@ def generate_autoyast(self, profile=None, system=None, raw_data=None):
...
 
     def generate_repo_stanza(self, obj, is_profile=True):
-
         """
-        Automatically attaches yum repos to profiles/systems in automatic
-        installation files (kickstart files) that contain the magic
-        $yum_repo_stanza variable.  This includes repo objects as well as the
-        yum repos that are part of split tree installs, whose data is stored
-        with the distro (example: RHEL5 imports)
+        Automatically attaches yum repos to profiles/systems in automatic installation files (kickstart files) that
```
Comment (hbokh) [inline cobbler/autoinstallgen.py:176]: "kickstart" --> "template"

File: cobbler/autoinstallgen.py [hunk 18]
```diff
@@ -210,11 +237,13 @@ def generate_repo_stanza(self, obj, is_profile=True):
         return buf
 
     def generate_config_stanza(self, obj, is_profile=True):
-
         """
-        Add in automatic to configure /etc/yum.repos.d on the remote system
-        if the automatic installation file (kickstart file) contains the magic
-        $yum_config_stanza.
+        Add in automatic to configure /etc/yum.repos.d on the remote system if the automatic installation file
+        (kickstart file) contains the magic $yum_config_stanza.
```
Comment (hbokh) [inline cobbler/autoinstallgen.py:242]: "kickstart" --> "template"

File: cobbler/clogger.py [hunk 19]
```diff
@@ -25,9 +25,9 @@
 import logging.config
 import os
 
-# Temporary hack, a clean solution seems to be tricky
-# Defining a variable in our Apache startup code seem not to
-# it is still set later when this code is executed via cobbler
+# Temporary hack, a clean solution seems to be tricky.
+# Defining a variable in our Apache startup code seem not to it is still set later when this code is executed via
```
Comment (hbokh) [inline cobbler/clogger.py:29]: Add a comma (,) between "to" and "it"

File: cobbler/cobblerd.py [hunk 20]
```diff
@@ -32,6 +32,11 @@
 
 
 def core(api):
+    """
+    Starts cobbler.
```
Comment (hbokh) [inline cobbler/cobblerd.py:36]: Nitpick: I notice both "cobbler" and "Cobbler" are used in various places in all files.

File: cobbler/download_manager.py [hunk 21]
```diff
@@ -39,13 +41,29 @@ def __init__(self, collection_mgr, logger=None):
             self.proxies = {}
 
     def urlread(self, url, proxies=None, cert=None):
+        """
+        Read the content of a given URL and pass the requests.Response object to the caller.
```
Comment (hbokh) [inline cobbler/download_manager.py:45]: Add a space between the two sentences.

File: cobbler/items/file.py [hunk 22]
```diff
@@ -60,13 +60,22 @@ class File(resource.Resource):
     #
 
     def make_clone(self):
+        """
+        Clone this file object. Please manually adjust all value yourself to make the cloned object unique.
```
Comment (hbokh) [inline cobbler/items/file.py:64]: Fix typo: value --> values

File: cobbler/items/file.py [hunk 23]
```diff
@@ -60,13 +60,22 @@ class File(resource.Resource):
...
+        :return: The cloned instance of this object.
+        """
         _dict = self.to_dict()
         cloned = File(self.collection_mgr)
         cloned.from_dict(_dict)
         return cloned
 
     def get_fields(self):
+        """
+        Return all fields which this class has with it's current values.
```
Comment (hbokh) [inline cobbler/items/file.py:75]: Fix typo: "it's" --> "its".

See https://grammarist.com/spelling/its-its/

File: cobbler/items/image.py [hunk 24]
```diff
@@ -72,13 +72,22 @@ class Image(item.Item):
...
+        :return: The cloned instance of this object.
+        """
         _dict = self.to_dict()
         cloned = Image(self.collection_mgr)
         cloned.from_dict(_dict)
         return cloned
 
     def get_fields(self):
+        """
+        Return all fields which this class has with it's current values.
```
Comment (hbokh) [inline cobbler/items/image.py:87]: Fix typo: "it's" --> "its".

File: cobbler/items/profile.py [hunk 25]
```diff
@@ -93,14 +93,21 @@ def __init__(self, *args, **kwargs):
...
+        """
         _dict = self.to_dict()
         cloned = Profile(self.collection_mgr)
         cloned.from_dict(_dict)
         return cloned
 
     def get_fields(self):
         """
-        Return the list of fields and their properties
+        Return all fields which this class has with it's current values.
```
Comment (hbokh) [inline cobbler/items/profile.py:108]: Fix typo: "it's" --> "its".

File: cobbler/items/package.py [hunk 26]
```diff
@@ -52,15 +52,28 @@ class Package(resource.Resource):
...
+        :return: The cloned instance of this object.
+        """
         _dict = self.to_dict()
         cloned = Package(self.collection_mgr)
         cloned.from_dict(_dict)
         return cloned
 
     def get_fields(self):
+        """
+        Return all fields which this class has with it's current values.
```
Comment (hbokh) [inline cobbler/items/package.py:67]: Fix typo: "it's" --> "its".

File: cobbler/items/repo.py [hunk 27]
```diff
@@ -115,6 +129,8 @@ def set_mirror(self, mirror):
         """
         A repo is (initially, as in right now) is something that can be rsynced.
         reposync/repotrack integration over HTTP might come later.
+
+        :param mirror: The mirroruri.
```
Comment (hbokh) [inline cobbler/items/repo.py:133]: Maybe better use "The mirror URI"?

File: cobbler/modules/authentication/ldap.py [hunk 28]
```diff
@@ -34,16 +33,25 @@
...
+    :rtype: str
     """
 
     return "authn"
 
 
 def authenticate(api_handle, username, password):
     """
-    Validate an ldap bind, returning True/False
+    Validate an ldap bind, returning whether the authentication was successful or not.
```
Comment (hbokh) [inline cobbler/modules/authentication/ldap.py:46]: Consistency: use LDAP, as on L99 in this file.

File: cobbler/modules/managers/bind.py [hunk 29]
```diff
@@ -391,6 +401,10 @@ def __write_secondary_conf(self):
     def __ip_sort(self, ips):
         """
         Sorts IP addresses (or partial addresses) in a numerical fashion per-octet or quartet
+
+        :param ips: A list of all IP (v6 and v4 mixed possible) which shall be sorted.
```
Comment (hbokh) [inline cobbler/modules/managers/bind.py:405]: "IP" --> "IP addresses", just as two lines below here.

File: cobbler/modules/managers/bind.py [hunk 30]
```diff
@@ -606,4 +629,11 @@ def write_dns_files(self):
 
 
 def get_manager(collection_mgr, logger):
+    """
+    This returns the object to manage a bind server located locally on the cobbler server.
```
Comment (hbokh) [inline cobbler/modules/managers/bind.py:633]: Nitpick: "BIND", just as on L623 in this file.

File: cobbler/modules/managers/dnsmasq.py [hunk 31]
```diff
@@ -205,4 +241,11 @@ def sync_dhcp(self):
 
 
 def get_manager(collection_mgr, logger):
+    """
+    Creates a manager object to manage an dnsmasq server.
```
Comment (hbokh) [inline cobbler/modules/managers/dnsmasq.py:245]: Fix typo: "a dnsmasq server", not "an dnsmasq server".

**Comment by SchoolGuy** [comment]
So I now have rewritten the git history as clean as I can get it. If @hbokh agrees I would merge.

**Comment by SchoolGuy** [comment]
All changes of @hbokh were addressed and no new comments from other maintainers were received. I will merge this after CI passes. For new problems about the docs please open PRs and/or issues.

File: cobbler/templar.py [hunk 32]
```diff
@@ -218,11 +224,11 @@ def render_cheetah(self, raw_data, search_table, subject=None):
...
-        data_input is either a string or a filename
-        search_table is a dict of metadata keys and values
-        out_path if not-none writes the results to a file
-        (though results are always returned)
-        subject is a profile or system object, if available (for snippet eval)
+
+        :param raw_data: Is the template code which is not rendered into the result.
+        :param search_table: is a dict of metadata keys and values
+        :param subject: is a profile or system object, if available (for snippet eval)
+        :return:
```
Comment (watologo1) [inline cobbler/templar.py:231]: Remove :return:

File: cobbler/tftpgen.py [hunk 33]
```diff
@@ -130,7 +142,14 @@ def copy_single_image_files(self, img):
         utils.linkfile(filename, newfile, api=self.api, logger=self.logger)
 
     def write_all_system_files(self, system, menu_items):
+        """
+        Writes all files for tftp for a given system with the menu items handed to this method. The system must have a
+        profile attached. Otherwise this method throws an error.
 
+        :param system: The system to generate files for.
+        :param menu_items:
+        :return:
```
Comment (watologo1) [inline cobbler/tftpgen.py:151]: Remove  unfilled docstring items :param menu_items: and :return:

File: cobbler/tftpgen.py [hunk 34]
```diff
@@ -306,6 +325,11 @@ def get_menu_items(self, arch=None):
         """
         Generates menu items for pxe and grub
         grub menu items are grouped into submenues by profie
+
```
Comment (watologo1) [inline cobbler/tftpgen.py:328]: Also correct above docstring method description:
- One line
- submenues -> submenus
- profie -> profile

File: cobbler/tftpgen.py [hunk 35]
```diff
@@ -367,11 +391,27 @@ def write_pxe_file(self, filename, system, profile, distro, arch,
...
-        that would appear inside the system object in api.py
 
+        More system-specific configuration may come in later, if so that would appear inside the system object in api.py
         Can be used for different formats, "pxe" (default) and "grub".
+
+        :param filename: If present this writes the output into the giving filename. If not present this method just
+                         returns the generated configuration.
+        :param system: TODO
+        :param profile: TODO
+        :param distro: TODO
```
Comment (watologo1) [inline cobbler/tftpgen.py:402]: Do not include TODOs here, better remove the entries.

File: cobbler/tftpgen.py [hunk 36]
```diff
@@ -580,10 +617,20 @@ def write_pxe_file(self, filename, system, profile, distro, arch,
...
-    def build_kernel_options(self, system, profile, distro, image, arch,
-                             autoinstall_path):
+    def build_kernel_options(self, system, profile, distro, image, arch, autoinstall_path):
         """
         Builds the full kernel options line.
+
+        :param system: TODO
+        :param profile: TODO
+        :param distro: TODO
+        :param image: TODO
```
Comment (watologo1) [inline cobbler/tftpgen.py:627]: Remove TODO entries.

File: cobbler/tftpgen.py [hunk 37]
```diff
@@ -878,6 +920,16 @@ def write_templates(self, obj, write_file=False, path=None):
         return results
 
     def generate_gpxe(self, what, name):
+        """
+        Generate the gpxe files.
+
+        :param what: either "profile" or "system". All other item types not valdi.
```
Comment (watologo1) [inline cobbler/tftpgen.py:926]: not valdi --> are not valid

File: cobbler/utils.py [hunk 38]
```diff
@@ -91,12 +91,18 @@ def _(foo):
 
 
 def die(logger, msg):
+    """
+    This method arifically let's Cobbler crash with an exception. Log the exception once in the per-task log or the main
```
Comment (watologo1) [inline cobbler/utils.py:95]: arificially -> wrong spelled, better remove the word totally.

Below: Remove unfilled docstrings

File: cobbler/utils.py [hunk 39]
```diff
@@ -112,6 +118,8 @@ def die(logger, msg):
 def log_exc(logger):
     """
     Log an exception.
+
+    :param logger:
```
Comment (watologo1) [inline cobbler/utils.py:122]: Here and next hunk: Remove unfilled docstrings

File: cobbler/utils.py [hunk 40]
```diff
@@ -156,6 +183,10 @@ def pretty_hex(ip, length=8):
 def get_host_ip(ip, shorten=True):
     """
     Return the IP encoding needed for the TFTP boot tree.
+
+    :param ip: The IP address to pretty print.
+    :param shorten: Wheather the IP-Address should be shortened or not.
```
Comment (watologo1) [inline cobbler/utils.py:188]: Wheather -> Whether

File: cobbler/utils.py [hunk 41]
```diff
@@ -189,6 +220,9 @@ def _IP(ip):
 def is_ip(strdata):
     """
     Return whether the argument is an IP address.
+
+    :param strdata: str
```
Comment (watologo1) [inline cobbler/utils.py:224]: :param strdata: str
-> :type strdata: str
?

File: cobbler/utils.py [hunk 42]
```diff
@@ -366,14 +426,15 @@ def find_initrd(path):
...
-    or as a URI.
+    Reads the contents of a file, which could be referenced locally or as a URI.
 
-    Returns None if file is remote and templating of remote files is
-    disabled.
-
-    Throws a FileNotFoundException if the file does not exist at the
-    specified location.
+    :param file_location: The location of the file to read.
+    :param logger: The logger to autdit this action with.
```
Comment (watologo1) [inline cobbler/utils.py:432]: autdit -> audit

File: cobbler/utils.py [hunk 43]
```diff
@@ -504,6 +581,14 @@ def input_string_or_dict(options, allow_multiples=True):
 
 
 def input_boolean(value):
+    """
+    Convert a str to a boolean. If this is not possible or the value is false return false.
+
+    :param value: The value to convet to boolean.
```
Comment (watologo1) [inline cobbler/utils.py:587]: convet -> convert

File: cobbler/utils.py [hunk 44]
```diff
@@ -512,6 +597,13 @@ def input_boolean(value):
 
 
 def update_settings_file(data):
+    """
+    Write data handed to this function into the settings file of Cobbler. This function overwrites the existing content.
+
+    :param data: The data to put into the settings file.
+    :return: True if the action succeded. Otherwise return nothing.
```
Comment (watologo1) [inline cobbler/utils.py:604]: succeded -> succeeded

File: cobbler/utils.py [hunk 45]
```diff
@@ -611,9 +709,16 @@ def blender(api_handle, remove_dicts, root_obj):
...
 
 def flatten(data):
-    # convert certain nested dicts to strings.
-    # this is only really done for the ones koan needs as strings
-    # this should not be done for everything
+    """
+    Convert certain nested dicts to strings. This is only really done for the ones koan needs as strings this should
+    not be done for everything
+
+    :param data: The dictionary where various keys should be converted into a string.
```
Comment (watologo1) [inline cobbler/utils.py:716]: where  -> in which
?

File: cobbler/utils.py [hunk 46]
```diff
@@ -657,14 +770,16 @@ def uniquify(seq):
...
 def __consolidate(node, results):
     """
-    Merge data from a given node with the aggregate of all
-    data from past scanned nodes.  Dictionaries and arrays are treated
-    specially.
+    Merge data from a given node with the aggregate of all data from past scanned nodes. Dictionaries and arrays are
+    treated specially.
+
+    :param node: The object to merge data into. The data from the node always wins.
+    :param results: The results dictionary.
```
Comment (watologo1) [inline cobbler/utils.py:777]: results -> resulting
Maybe better: Merged data as dictionary
?

File: cobbler/utils.py [hunk 47]
```diff
@@ -762,8 +884,15 @@ def dict_to_string(_dict):
...
-    by args are appended to the command
+    Sync files from src to dst. The extra arguments specified by args are appended to the command.
+
+    :param src: The source for the copy process.
+    :param dst: The destionation for the copy process.
+    :param args: The extra arguments are appended to our standard arguements.
+    :param logger: The logger to audit the action with.
+    :param quiet: If "True" no progress is reported. If "False" then progress will be reported by rsync.
+    :type quiet: bool
+    :return:
```
Comment (watologo1) [inline cobbler/utils.py:895]: Remove empty :return:

File: cobbler/utils.py [hunk 48]
```diff
@@ -1060,6 +1227,14 @@ def linkfile(src, dst, symlink_ok=False, cache=True, api=None, logger=None):
 
 
 def copyfile(src, dst, api=None, logger=None):
+    """
+    Copy a file from source to the destination.
+
+    :param src: The source file. This may also be a folder.
+    :param dst: The destion for the file or folder.
```
Comment (watologo1) [inline cobbler/utils.py:1234]: destion -> destination

File: cobbler/utils.py [hunk 49]
```diff
@@ -1416,13 +1714,24 @@ def set_virt_cpus(self, num):
...
         return
     if var.find("..") != -1 or var.find(";") != -1:
         raise CX("Invalid characters found in input")
 
 
 def is_selinux_enabled():
+    """
+    This check is achieved via a subprocess call to ``selinuxenabled``. Default return is false.
+
+    :return: Weather selinux is enabled or not.
```
Comment (watologo1) [inline cobbler/utils.py:1732]: Weather -> Whether

File: cobbler/utils.py [hunk 50]
```diff
@@ -1448,21 +1756,50 @@ class MntEntObj(object):
...
+        """
         return "%s %s %s %s %s %s" % (self.mnt_fsname, self.mnt_dir, self.mnt_type,
                                       self.mnt_opts, self.mnt_freq, self.mnt_passno)
 
 
 def get_mtab(mtab="/etc/mtab", vfstype=None):
+    """
+    Get the list of mtab entries. If a custum mtab should be read then the location can be overriden via a parameter.
+
+    :param mtab: The location of the mtab. Arguemnt can be ommited if the mtab is at its default location.
```
Comment (watologo1) [inline cobbler/utils.py:1796]: Arguemnt -> Argument

File: cobbler/utils.py [hunk 51]
```diff
@@ -1509,6 +1865,12 @@ def set_serial_baud_rate(self, baud_rate):
 
 
 def __cache_mtab__(mtab="/etc/mtab"):
+    """
+    Open the mtab and cache it inside Cobbler. If it is guessed that the mtab hasn't changed the cache data is used.
+
+    :param mtab: The location of the mtab. Arguemnt can be ommited if the mtab is at its default location.
```
Comment (watologo1) [inline cobbler/utils.py:1871]: Arguemnt -> Argument

File: cobbler/utils.py [hunk 52]
```diff
@@ -1517,14 +1879,18 @@ def __cache_mtab__(mtab="/etc/mtab"):
...
+        - the device the file is on
+        - the path of the file relative to the device.
+    For example:
          /boot/vmlinuz -> (/dev/sda3, /vmlinuz)
          /boot/efi/efi/redhat/elilo.conf -> (/dev/cciss0, /elilo.conf)
          /etc/fstab -> (/dev/sda4, /etc/fstab)
-    '''
+
+    :param fname: The filename to split up.
+    :return: A touple containing the device and relative filename.
```
Comment (watologo1) [inline cobbler/utils.py:1892]: touple -> tuple

File: cobbler/utils.py [hunk 53]
```diff
@@ -1573,6 +1946,16 @@ def is_remote_file(file):
...
 def subprocess_sp(logger, cmd, shell=True, input=None):
+    """
+    Call a shell process and redirect the output for internal usage.
+
+    :param logger: The logger to audit the action with.
+    :param cmd: The command to execute in a subprocess call.
+    :param shell: Whether to use a shell or not for the execution of the commmand.
+    :type shell: bool
+    :param input: If there is any input needed for that command to stdin.
+    :return: A touple of the output and the return code.
```
Comment (watologo1) [inline cobbler/utils.py:1957]: touple -> tuple

File: cobbler/utils.py [hunk 54]
```diff
@@ -1597,20 +1980,54 @@ def subprocess_sp(logger, cmd, shell=True, input=None):
...
+    """
+    A simple subprocess call with no return code capturing.
+
+    :param logger: The logger to audit the action with.
+    :param cmd: The command to execute.
+    :param shell: Whether to use a shell or not for the execution of the commmand.
+    :type shell: bool
+    :param input: If there is any input needed for that command to stdin.
+    :return: The data which the subprocess returns.
+    """
```
Comment (watologo1) [inline cobbler/utils.py:2007]: docstring 3 times longer than the method and C&P from above...
Better add a reference to the other method or leave docstrings outs

File: cobbler/utils.py [hunk 55]
```diff
@@ -1655,6 +2077,13 @@ def clear_from_fields(item, fields, is_subobject=False):
 
 
 def from_dict_from_fields(item, item_dict, fields):
+    """
+    Not known what this method does exactly.
+
+    :param item: Not known what this is needed for exactly.
+    :param item_dict: Not known what this is needed for exactly.
+    :param fields: Not known what this is needed for exactly.
```
Comment (watologo1) [inline cobbler/utils.py:2085]: Better remove docstrings if not filled.

File: cobbler/utils.py [hunk 56]
```diff
@@ -1689,15 +2118,22 @@ def from_dict_from_fields(item, item_dict, fields):
...
 
 def to_dict_from_fields(item, fields):
+    """
+    Not known what this method does exactly.
+
+    :param item: Not known what this is needed for exactly.
+    :param fields: Not known what this is needed for exactly.
+    :return: Not known what this is returning
+    :rtype: dict
+    """
```
Comment (watologo1) [inline cobbler/utils.py:2128]: Returns a dictionary of the fields of an item (distro, profile,..).

File: cobbler/utils.py [hunk 57]
```diff
@@ -1711,6 +2147,12 @@ def to_dict_from_fields(item, fields):
 def to_string_from_fields(item_dict, fields, interface_fields=None):
     """
     item_dict is a dictionary, fields is something like item_distro.FIELDS
+
+    :param item_dict: Not known what this is needed for exactly.
+    :param fields: Not known what this is needed for exactly.
+    :param interface_fields: Not known what this is needed for exactly.
+    :return: Not known what this is returning exactly.
+    :rtype: str
```
Comment (watologo1) [inline cobbler/utils.py:2155]: Better remove unfilled docs here.

File: cobbler/utils.py [hunk 58]
```diff
@@ -1948,19 +2437,40 @@ def lod_to_dod(_list, indexkey):
 
 
 def mycmp(x, y):
+    """
+    Compares if x is smaller than y.
+
+    :param x: The first parameter.
+    :param y: The second parameter.
+    :return: The bool of the action.
+    :rtype: bool
```
Comment (watologo1) [inline cobbler/utils.py:2446]: ?!? This doesn't tell anythiing, better remove param descriptions.

File: cobbler/utils.py [hunk 59]
```diff
@@ -1985,6 +2501,13 @@ def namedconf_location(api):
 
 
 def zonefile_base(api):
+    """
+    This determins the base directory for the zone files which are important for the named service which Cobbler tries
```
Comment (watologo1) [inline cobbler/utils.py:2505]: determins -> determines

File: cobbler/utils.py [hunk 60]
```diff
@@ -2040,24 +2583,46 @@ def link_distro(settings, distro):
 
 
 def find_distro_path(settings, distro):
+    """
+    This returns the absolute path to the distro under the ``distro_mirror`` directory. If that directory doesn't
+    contain the kernel, the directory of the kernel in the distro is returned.
+
+    :param settings: The settings to resolve user configurable actions with.
+    :param distro: The distroibution to find the path of.
```
Comment (watologo1) [inline cobbler/utils.py:2591]: distroibution -> distribution

File: cobbler/validate.py [hunk 61]
```diff
@@ -40,9 +40,12 @@ def object_name(name, parent):
     """
     Validate the object name.
 
-    @param: str name (object name)
-    @param: str parent (parent object name)
-    @returns: str name or CX
+    :param name: object name
+    :type name: str
+    :param parent: (parent object name)
```
Comment (watologo1) [inline cobbler/validate.py:45]: Also remove brackets here.

File: cobbler/validate.py [hunk 62]
```diff
@@ -210,8 +223,11 @@ def name_servers_search(search, for_item=True):
     """
     Validate nameservers search domains.
 
-    @param: str/list search (string or list of search domains)
-    @param: bool for_item (enable/disable special handling for Item objects)
+    :param search: One more more search domains to validate.
```
Comment (watologo1) [inline cobbler/validate.py:226]: One more more -> One or more

File: cobbler/actions/check.py [hunk 63]
```diff
@@ -311,7 +378,9 @@ def check_tftpd_dir(self, status):
 
     def check_ctftpd_dir(self, status):
         """
-        Check if cobbler.conf's tftpboot directory exists
+        Check if ``cobbler.conf``'s tftpboot directory exists.
+
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply cobbler/actions/check.py:384]: @watologo1 In the `rst` syntax a double grave accent has a different meaning then you have assumed: http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup

**Pull request merged**